mitiq/qse/qse_utils.py

```diff
@@ -53,82 +54,58 @@ def get_expectation_value_for_observable(
     circuit: QPROGRAM,
     executor: Union[Executor, Callable[[QPROGRAM], QuantumResult]],
     observable: Union[PauliString, Observable],
-    pauli_string_to_expectation_cache: Dict[PauliString, complex] = {},
+    pauli_expectation_cache: Dict[PauliString, complex] = {},
 ) -> float:
     """Provide pauli_string_to_expectation_cache if you want to take advantage
     of caching.
 
     This function modifies pauli_string_to_expectation_cache in place.
     """
 
+    final_executor = (
+        executor if isinstance(executor, Executor) else Executor(executor)
+    )
+
     def get_expectation_value_for_one_pauli(
         pauli_string: PauliString,
     ) -> float:
         cache_key = pauli_string.with_coeff(1)
-        pauli_string_to_expectation_cache[cache_key] = final_executor.evaluate(
+        pauli_expectation_cache[cache_key] = final_executor.evaluate(
             circuit, Observable(cache_key)
         )[0]
-        return (
-            pauli_string_to_expectation_cache[cache_key] * pauli_string.coeff
-        ).real
+        return (pauli_expectation_cache[cache_key] * pauli_string.coeff).real
 
-    total_expectation_value_for_observable = 0.0
-    final_executor: Executor = (
-        executor if isinstance(executor, Executor) else Executor(executor)
+    paulis = (
+        [observable]
+        if isinstance(observable, PauliString)
+        else observable.paulis
     )
-
-    if isinstance(observable, PauliString):
-        pauli_string = observable
-        total_expectation_value_for_observable += (
-            get_expectation_value_for_one_pauli(pauli_string)
-        )
-    elif isinstance(observable, Observable):
-        for pauli_string in observable.paulis:
-            total_expectation_value_for_observable += (
-                get_expectation_value_for_one_pauli(pauli_string)
-            )
-    return total_expectation_value_for_observable
+    expectation_value = sum(
+        get_expectation_value_for_one_pauli(pauli) for pauli in paulis
+    )
+    return expectation_value
 
 
 def _compute_overlap_matrix(
     circuit: QPROGRAM,
     executor: Union[Executor, Callable[[QPROGRAM], QuantumResult]],
     check_operators: Sequence[PauliString],
-    pauli_string_to_expectation_cache: Dict[PauliString, complex] = {},
+    pauli_expectation_cache: Dict[PauliString, complex] = {},
+    code_hamiltonian: Optional[Observable] = None,
```

Comment:
Is this new argument here in order to get rid of `_compute_hamiltonian_overlap_matrix` ?
Seems good, just trying to understand.

Reply:
Yup that's exactly right! We previously had

1. `_compute_hamiltonian_overlap_matrix`, and
2. `_compute_overlap_matrix`

which were largely the same code, except for one line. I combined the two functions in order to cut down on repeated code.